### PR TITLE
Added Visualisation File type/widget/formatter

### PR DESCRIPTION
**Pull request by jez500**
Hi @tobybellwood & @jorgegc 

Can you please have a look at this one. The main change is DVF now supports file uploads rather than just a remote URL. This works really well with CSVs but will support anything really given the correct data  source.

Other Changes:
* Readme added! with some basic "how to use" instructions
* A number of bug fixes 
* Lots of coding standard fixes
* composer file added

I am keen to get this into my project asap and noted github has diverged a bit from drupal.org, happy to clone it in manually but ideally would be great to see drupal.org project updated with the most recent code.

Cheers

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+entity. Sub modules can provide data sources or visualisation plugins that use 
+this field type, see related modules above for modules currently available.
+
+## Installing & Configuration
+
+* Install DVF and a suitable data source (eg. `dvf_csv`) then enable.
+* Then visit `Manage fields` for the entity type you want to add the 
+  visualisation to eg `/admin/structure/types/manage/page/fields`
+* Click `Add field` and select either `Visualisation URL` or 
+  `Visualisation File` depeing on if the data source will be local or remote,
```
Comment (jorgegc) [inline README.md]: ~~depeing~~ depending

File: src/FieldFormatterTrait.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+   *   The language that should be used to render the field.
+   *
+   * @return array
+   *   A renderable array for $items, as an array of child elements keyed by
+   *   consecutive numeric indexes starting from 0.
+   */
+  public function viewElements(FieldItemListInterface $items, $langcode) {
+    $element = [];
+
+    /** @var \Drupal\dvf\Plugin\Field\FieldType\VisualisationUrlItem $item */
```
Comment (jorgegc) [inline src/FieldFormatterTrait.php]: We can't type hint this as `\Drupal\dvf\Plugin\Field\FieldType\VisualisationUrlItem`. Ideally, these fields should share a common interface `VisualisationItem` which guarantees `VisualisationItem::getVisualisationPlugin` exists.

File: src/Plugin/Field/FieldWidget/VisualisationFileWidget.php [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+  /**
+   * {@inheritdoc}
+   */
+  public function massageFormValues(array $values, array $form, FormStateInterface $form_state) {
+    // See FileWidget::massageFormValues() for the main reason for this
+    // method. It also restructures the values to suit the database schema.
+    $new_values = [];
+    $file_defaults = ['display' => 1, 'description' => ''];
+    foreach ($values as &$value) {
```
Comment (jorgegc) [inline src/Plugin/Field/FieldWidget/VisualisationFileWidget.php]: Line break before control structures improves code readability.

File: src/Plugin/Field/FieldWidget/VisualisationFileWidget.php [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+          $new_values[] = $new_value;
+        }
+      }
+    }
+
+    return !empty($new_values) ? $new_values : $values;
+  }
+
+  /**
+   * Form API callback. Retrieves the value for the file_generic field element.
```
Comment (jorgegc) [inline src/Plugin/Field/FieldWidget/VisualisationFileWidget.php]: `Form API callback: Retrieves the value for the file_generic field element.`?

**Review by jorgegc — CHANGES_REQUESTED** [review]
Hey @jez500 very nice work!

There is one thing that I think we should consider adding to this PR before merging, and that is the common interface for Visualisation field types.

Please let me know your thoughts.

Cheers

**Comment by jez500** [comment]
Thanks for the feedback @jorgegc those changes have been addressed

FYI @tobybellwood

**Review by jorgegc — APPROVED** [review]

**Comment by tobybellwood** [comment]
Cool - I'll merge here and push up to d.o. - the only difference up to now was the bundling of the submodules in the Github version wasn't reflected in d.o.

**Comment by tobybellwood** [comment]
d.o. updated https://www.drupal.org/project/dvf/releases/8.x-1.x-dev (and updated dvf_ckan and dvf_csv d.o. pages to point to submodules of dvf instead)

**Comment by jez500** [comment]
Thanks @tobybellwood :)

**Pull request merged**